repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java

```diff
@@ -472,9 +483,11 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
             outLineageInfo.setHasMoreOutputs(true);
             hasRelationsLimitReached = true;
         } else {
-            if (! (direction.equals(AtlasLineageOnDemandInfo.LineageDirection.INPUT) && outGuid.equals(baseGuid))) {
-                outLineageInfo.incrementOutputRelationsCount();
-            }
+            outLineageInfo.incrementOutputRelationsCount();
+        }
+
+        if (lineageContainsVisitedEdge) {
+            return false;
```

Comment:
if we keep this as false, this method won't return true if the limit is reached. need to update this line as:

`return hasRelationsLimitReached;`